src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs

```diff
@@ -108,30 +110,43 @@ internal static class NamedPipeUtils
                 // The starttime is there to prevent another process easily guessing the pipe name
                 // and squatting on it.
                 // There is a limit of 104 characters in total including the temp path to the named pipe file
-                // on non-Windows systems, so we'll convert the starttime to hex and just take the first 8 characters.
+                // on non-Windows systems, so we'll convert the starttime to hex and just take 8 characters.
 #if UNIX
                 .Append(proc.StartTime.ToFileTime().ToString("X8").AsSpan(1, 8))
 #else
                 .Append(proc.StartTime.ToFileTime().ToString(CultureInfo.InvariantCulture))
 #endif
                 .Append('.')
-                .Append(proc.Id.ToString(CultureInfo.InvariantCulture))
-                .Append('.')
+                .Append(proc.Id.ToString(CultureInfo.InvariantCulture));
+
+            int uniqueLength = pipeNameBuilder.Length;
+            pipeNameBuilder.Append('.')
                 .Append(CleanAppDomainNameForPipeName(appDomainName))
                 .Append('.')
                 .Append(proc.ProcessName);
+
+            string pipeName = pipeNameBuilder.ToString();
 #if UNIX
-            int charsToTrim = pipeNameBuilder.Length - MaxNamedPipeNameSize;
+            // .NET Creates named pipes in the tmp path with a special prefix which needs to be included in the max
+            // length calculation.
+            string pipePrefix = Path.Combine(Path.GetTempPath(), "CoreFxPipe_");
+            int remainingPipeLength = Math.Max(MaxNamedPipeNameSize - pipePrefix.Length, 0);
+
+            int charsToTrim = pipeName.Length - remainingPipeLength;
             if (charsToTrim > 0)
             {
                 // TODO: In the case the pipe name is truncated, the user cannot connect to it using the cmdlet
                 // unless we add a `-Force` type switch as it attempts to validate the current process name
                 // matches the process name in the pipe name
-                pipeNameBuilder.Remove(MaxNamedPipeNameSize + 1, charsToTrim);
+                pipeName = pipeName.Substring(0, Math.Max(pipeName.Length - charsToTrim, 0));
+                if (pipeName.Length < uniqueLength)
+                {
+                    throw new PSArgumentNullException(nameof(proc));
```

Comment:
We should throw a better error and message here, since this isn't really an argument error but an operational limitation.

Reply:
That's a good idea, I wasn't really too sure what exception to use at the time.